### PR TITLE
Hander for default image loading

### PR DESCRIPTION
**Pull request by aquatiko**
Added a utility function for handling the default Image loading.

Closes: #8

### CODE REVIEW

**Comment by aquatiko** [comment]
Also, an error is raised sometimes if the dimensions of Image being loaded is not appropriate.
Eg: [euve.fits](https://fits.gsfc.nasa.gov/samples/EUVEngc4151imgx.fits) , although the first HUD is image type, error raises due to inappropriate image dimensions. 
Ideas to handle that are welcome, so that I can put it along with this patch.

**Comment by giordano** [comment]
I don't like the approach, you're opening the file several times, you could at least re-use it.  You should be able to open it once.
Also, not that in general
```julia
typeof(x) == T
```
can be replaced with
```julia
x isa T
```
In addition, if you're testing types, you're probably better off defining methods with different signatures, i.e.
```julia
if x isa T
    # do something
else
    # do something else
end
```
in Julia is more naturally expressed with
```julia
f(x::T) = # do something
f(x) = # do something else
```

Remember also that I (and *many* people like me) will never accept a PR without tests, unless there is a great reason not to add them.  Tests are what ensure that a package is always working as expected, if you don't add them there will be nothing in the future that will tell you that you didn't unintentionally break anything.


> Also, an error is raised sometimes if the dimensions of Image being loaded is not appropriate.
Eg: [euve.fits](https://fits.gsfc.nasa.gov/samples/EUVEngc4151imgx.fits) , although the first HUD is image type, error raises due to inappropriate image dimensions. 
Ideas to handle that are welcome, so that I can put it along with this patch. 

I think this problem is more connected to `FITSIO.jl` rather than this package.  Note that Astropy reports this table:
```
Filename: EUVEngc4151imgx.fits
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU      62   ()      
  1  ds            1 ImageHDU        72   (512, 512)   int16   
  2  sw_night      1 ImageHDU        78   (2048, 300)   int16   
  3  mw            1 ImageHDU        78   (2048, 300)   int16   
  4  lw            1 ImageHDU        78   (2048, 300)   int16   
  5  ds_limits     1 BinTableHDU     71   3R x 3C   [8A, 1E, 1E]   
  6  sw_night_limits    1 BinTableHDU     71   2R x 3C   [12A, 1E, 1E]   
  7  mw_limits     1 BinTableHDU     71   2R x 3C   [12A, 1E, 1E]   
  8  lw_limits     1 BinTableHDU     71   2R x 3C   [12A, 1E, 1E]
```

**Comment by aquatiko** [comment]
I haven't added the tests yet because I wanted some suggestions about the approach before adding total coverage. That's the approach I use regarding the PRs I make. 
In all, what points do you suggest for this PR?

**Comment by giordano** [comment]
> I haven't added the tests yet because I wanted some suggestions about the approach before adding total coverage. That's the approach I use regarding the PRs I make.

Ok, fine, but I wanted to stress the importance of tests :slightly_smiling_face:

Regarding tests, keep in mind that downloading FITS files from the internet is not a good idea because that would require always having an Internet connection to run tests.  Instead, we need to create the necessary FITS files.  This is a bit annoying but doesn't require Internet.

> In all, what points do you suggest for this PR?

I'd do something like this (not tested at all!): https://github.com/giordano/AstroImages.jl/commit/82f7f5c5fc3ffbf263c606a6c12fffd543a269d2  If I got it correctly, in #8 you're suggesting to fix the behaviour of the default extension.  I think that if one provides explicitly an extension should take care of selecting the right extension, if they provide a wrong extension they must expect an error, we don't have too smart in this case.

Thanks to multiple dispatch we can define a different method for the case where no extension is provided and a default value must be used.  In this case we can look in a smart way for the first ImageHDU.  In this case we have to open the file, and pass the result to the new `_load` function.

What do you think?

**Comment by aquatiko** [comment]
@giordano This seems great. Can we also add `@info "Image was loaded from HDU $ext"` in the code?
Also, should I add your code along with the tests supporting it in this PR?

**Comment by giordano** [comment]
> @giordano This seems great. Can we also add `@info "Image was loaded from HDU $ext"` in the code

I don't generally like being too verbose, but it's probably a good idea to tell the user which is the extension used, when the default extension is used.

> and some code to check if the `ImageHDU` have at least 2D data?

Yes, as suggested [here](https://github.com/JuliaAstro/FITSIO.jl/issues/111), this check should be sufficient to identify a real image.

> Also, can I add your code along with the tests supporting it in this PR?

Yes, please, go ahead!

File: src/AstroImages.jl [hunk 1]
```diff
@@ -2,10 +2,11 @@ __precompile__()
 
 module AstroImages
 
-using FITSIO, FileIO, Images
+using FITSIO, FileIO, Images, Interact
```
Comment (giordano) [inline src/AstroImages.jl]: Am I wrong or this isn't needed in this PR?

File: test/runtests.jl [hunk 2]
```diff
@@ -1,4 +1,4 @@
-using AstroImages, FITSIO, Images
+using AstroImages, FITSIO, Images, Random, Widgets
```
Comment (giordano) [inline test/runtests.jl]: Is `Widgets` needed here?

Comment (aquatiko) [reply test/runtests.jl]: Oh! These things slipped in from #9

**Comment by aquatiko** [comment]
@giordano The tests seems to fail in `appveyor` due to [this](https://ci.appveyor.com/project/giordano/astroimages-jl/builds/23569065/job/a9wbw51xiw4yyp0o#L278), would it be ok or do I need to fix it? Any pointers on how to fix that?

**Comment by giordano** [comment]
I'd say that it has to be fixed, even though it isn't clear to me at a first glance why it's failing :confused:

File: src/AstroImages.jl [hunk 3]
```diff
@@ -58,9 +59,28 @@ AstroImage(data::Matrix{T}) where {T<:Real} = AstroImage{T,Gray}(data)
...
+AstroImage(file::String, ext::Int) = AstroImage(Gray, file, ext)
+
+AstroImage(color::Type{<:Color}, fits::FITS, ext::Int) =
+    AstroImage(color, _load(fits, ext))
+function AstroImage(file::String)
+    fits = FITS(file)
+    ext = 0
+    for (i, hdu) in enumerate(fits)
+        if hdu isa ImageHDU && length(size(hdu)) >= 2	# check if Image is atleast 2D
+            ext = i
```
[the remark this answers is not in the document]
Comment (aquatiko) [reply src/AstroImages.jl]: Now I realised, it should be like this...
```
function AstroImage(file::String)
    fits = FITS(file)
    ext = 0
    for (i, hdu) in enumerate(fits)
        if hdu isa ImageHDU && length(size(hdu)) >= 2	# check if Image is atleast 2D
            ext = i
            break
        end
    end
    if ext > 1
       	@info "Image was loaded from HDU $ext"
    elseif ext == 0
        error("There are no ImageHDU extensions in \"$file\"")
    end
    AstroImage(Gray, fits, ext)
end
```
We want to show the info only when the default first HDU is not accessed.

**Comment by aquatiko** [comment]
I will remove the WIP tag once the PR is ready for review. I can't run appveyor locally, so I was testing it like this.

**Comment by giordano** [comment]
My comment was referring to the fact that you *never* want to do commits like [this](https://github.com/JuliaAstro/AstroImages.jl/commit/82f794eb5326b5ce4905f5c3999900bf59e07491).  If you want to take commits from another branch just merge the branch, or rebase on it if you know how to manage a rebase, but **don't** just copy the content of the file from another branch, you risk to screw the history and face merge conflicts that you wouldn't get otherwise.

I think that in #14 I'm going to fix part of the problems that we're facing on Windows with AppVeyor, but not of all them.  I've been wanting to fix the issue about non-closed files for a long time, but never did that actually.  After I'll merge that PR please don't do anything, I'll try to fix the history of this branch.

**Comment by aquatiko** [comment]
Actually I wasn't familiar that, like Travis we can make an account to test our fork branches appveyor. After I knew, I did several iterations and restructured the tests to work. I was going to push the code but was a little occupied with the proposal submission deadline😅
No worries if you don't want me to

**Comment by giordano** [comment]
If you have some changes now, please go ahead and push them!

File: test/runtests.jl [hunk 4]
```diff
@@ -1,8 +1,10 @@
-using AstroImages, FITSIO, Images
+using AstroImages, FITSIO, Images, Random
 using Test
 
 import AstroImages: _float, render
 
+fname = tempname() * ".fits"
```
Comment (giordano) [inline test/runtests.jl]: I'm confused, now this file is not removed at the end any more?

Comment (aquatiko) [reply test/runtests.jl]: I was suspicious of `rm` causing the `EBUSY` error on appvyeor. So I thought using it just once should be good enough. Fixed in the latest commit, seems to be working well that way too.

Comment (giordano) [reply test/runtests.jl]: Well, there was no `rm` at all in https://github.com/JuliaAstro/AstroImages.jl/blob/024a521f09cf590a34b7335605300b394a539346/test/runtests.jl :wink:

**Pull request merged**